Add openAPI specification

iamalexvybornyi opened the pull request.

iamalexvybornyi committed:
Add the first version of the openAPI specification

iamalexvybornyi committed:
Update one of the operation names (/passport/findByGivenDates -> /passport/findByGivenDateBetween), format document using IntelliJ IDEA

iamalexvybornyi committed:
Update api spec according to the comments

justYP commented on passport-management-system.yml:
```diff
@@ -0,0 +1,691 @@
+        enum:
+          - "Active"
+          - "Inactive"
+  ApiResponse:
```
Looks like it is used for error responses only. So I'd suggest another name for this object.

iamalexvybornyi replied:
Renamed to ApiError

justYP commented on passport-management-system.yml:
```diff
@@ -0,0 +1,691 @@
+        enum:
+          - "Active"
+          - "Inactive"
+  CreatePassportDto:
```
I'd use `Request`/`Response` instead of `Dto`

iamalexvybornyi replied:
Renamed accordingly (CreatePersonDto -> CreatePersonRequest, PersonDto -> PersonResponse, etc.)

justYP commented on passport-management-system.yml:
```diff
@@ -0,0 +1,691 @@
+        enum:
+          - "Internal"
+          - "International"
+      status:
```
is `status` required?

iamalexvybornyi replied:
Yes, added it to the list of required fields

justYP commented on passport-management-system.yml:
```diff
@@ -0,0 +1,691 @@
+        enum:
+          - "Internal"
+          - "International"
+      person:
```
Need to discuss

iamalexvybornyi replied:
Sure. I think this exact reference to the person entity will not lead to any issues as this is a PersonDto (now PersonResponse) and it doesn't have any references to passports. But the other entity (Person) has a reference to passports which might lead to the recursion. It seems like I will need to get rid of this reference but let's discuss it in detail later

iamalexvybornyi replied:
By the way, it seems like I can remove the "Person" and "Passport" models from the specification as they are not related to the API description, they are only related to the internal implementation details and are not referenced in the requests

justYP reviewed (commented):
Pls make sure the following requirements have covered:

- Get current active passports of a person 
- Search passports by given dates range

iamalexvybornyi committed:
Update specification according to the comments:
- Rename DTOs to Request/Response
- Rename ApiResponse to ApiError
- Add activeOnly query param for getting only active passports of a person
- Add status to the list of required fields

iamalexvybornyi commented:
> Pls make sure the following requirements have covered:
> 
> * Get current active passports of a person
> * Search passports by given dates range

Added the description of the functionality from the first item as I somehow missed it initially. The functionality from the second item has already been described (there are 2 query params in the /passports GET request (**startDate** and **endDate**). If both params are specified the request will return the passports between the given dates
